### PR TITLE
LLaMA2 weight loading

### PR DESCRIPTION
**Pull request by jmercat**
This is a different version of issue #36 that I have been working on in parallel. Sorry for the redundancy, I'll let the OpenLM team choose what they want to take.

Usage: 
Get LLaMA weights
Convert them using `python open_lm/utils/llama2openlm.py source_weights target_weights`
`cd eval`
`torchrun --nproc_per_node 3 python eval_openlm_ckpt.py --checkpoint ../checkpoints/llama2_7b.pt --model llama2_7b.json --eval-yaml in_memory_hf_eval.yaml --tokenizer <path_to_tokenizer>`

Here are the Jeopardy results from it:

~~metrics/jeopardy/10-shot/american_history/InContextLearningLMAccuracy: 0.5690072774887085
metrics/jeopardy/10-shot/literature/InContextLearningLMAccuracy: 0.6673469543457031
metrics/jeopardy/10-shot/science/InContextLearningLMAccuracy: 0.39915966987609863
metrics/jeopardy/10-shot/word_origins/InContextLearningLMAccuracy: 0.33424657583236694
metrics/jeopardy/10-shot/world_history/InContextLearningLMAccuracy: 0.6595174074172974~~

~~Average : 0.526~~

Edit: The previous results had a bug (the normalization layer parameter was not read from the config file but from argparse), new results are slightly better:

metrics/jeopardy/10-shot/american_history/InContextLearningLMAccuracy: 0.5738498568534851
metrics/jeopardy/10-shot/literature/InContextLearningLMAccuracy: 0.6693877577781677
metrics/jeopardy/10-shot/science/InContextLearningLMAccuracy: 0.4075630307197571
metrics/jeopardy/10-shot/word_origins/InContextLearningLMAccuracy: 0.34520548582077026
metrics/jeopardy/10-shot/world_history/InContextLearningLMAccuracy: 0.6568364500999451

Average: 0.530568516254425

Here is what I did:
- ~~Added script for weight conversion from llama to OpenLM (from https://gist.github.com/Vaishaal/b5367e9c0b6e5d59d9a7be2a4ba606a9)~~
- Added a model config for llama 7b
- Added a llama_rotary positional encoding (slightly modified from https://github.com/facebookresearch/llama/blob/main/llama/model.py to make it compatible with torch.jit)
- Modified eval_openlm_ckpt.py to accept a tokenizer input argument and load the weights with strict=False
- Changed the input flag for positional embedding selection to a input string argument with choices [rotary, llama_rotary, head_rotary]
- Added a torch.no_grad decorator to eval functions
- Fixed dependencies to utils.transformers.model that should have been utils.transformers.hf_model after previous filename change

### CODE REVIEW

File: eval/eval_openlm_ckpt.py [hunk 1]
```diff
@@ -104,11 +112,14 @@ def main():
...
-    tokenizer = GPTNeoXTokenizerFast.from_pretrained("EleutherAI/gpt-neox-20b")
+    if "gpt-neox-20b" in args.tokenizer:
+        tokenizer = GPTNeoXTokenizerFast.from_pretrained("EleutherAI/gpt-neox-20b")
+    elif "llama" in args.tokenizer:
+        tokenizer = LlamaTokenizerFast.from_pretrained(args.tokenizer)
 
     state_dict = checkpoint["state_dict"]
     state_dict = {x.replace("module.", ""): y for x, y in state_dict.items()}
-    open_lm.model.load_state_dict(state_dict)
+    open_lm.model.load_state_dict(state_dict, strict=False)
```
Comment (sagadre) [inline eval/eval_openlm_ckpt.py:122]: is this because llama has extra keys?

File: eval/in_memory_hf_eval.yaml [hunk 2]
```diff
@@ -30,9 +30,9 @@ fsdp_config:
 
 icl_tasks:
 -
-  label: mmlu
-  dataset_uri: local_data/mmlu.jsonl # ADD YOUR OWN DATASET URI
-  num_fewshot: [0]
-  icl_task_type: multiple_choice
-  continuation_delimiter: 'Answer: ' # this separates questions from answers
+  label: jeopardy
```
Comment (sagadre) [inline eval/in_memory_hf_eval.yaml:33]: any particular to switch to jeopardy here?

File: open_lm/model.py [hunk 3]
```diff
@@ -73,7 +74,7 @@ class Params:
     weight_tying: bool = False
     norm_type: nn.Module = nn.LayerNorm
     apply_qk_norm: bool = False
-    rotary_old: bool = False
+    positional_embedding_type: str = "rotary"
```
Comment (sagadre) [inline open_lm/model.py:77]: nice!

File: open_lm/model.py [hunk 4]
```diff
@@ -299,7 +309,7 @@ def create_params(args):
         weight_tying=cfg["weight_tying"],
         norm_type=get_norm_class(args.model_norm),
         apply_qk_norm=args.qk_norm,
-        rotary_old=args.rotary_old,
+        positional_embedding_type=cfg["positional_embedding_type"],
```
Comment (sagadre) [inline open_lm/model.py:312]: old configs don't have a "positional_embedding_type", can we set the default to be "rotary"?

File: open_lm/utils/llama2openlm.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
+"""
+This script converts the weights from LLAMA to OpenLM.
```
Comment (sagadre) [inline open_lm/utils/llama2openlm.py:2]: nit: maybe another open_lm/scripts folder?

Comment (achalddave) [reply open_lm/utils/llama2openlm.py:2]: (this one got merged in https://github.com/mlfoundations/open_lm/pull/45/files so will be removed from here)

File: README.md [hunk 6]
```diff
@@ -126,26 +126,26 @@ cd scripts
...
+Once this checkpoint has been downloaded, you can evaluate it by following the directions in the [Evaluate Model](#evaluate-model) section above and passing `--positional_embedding_type head_rotary` (see note below).
 
 Note: We trained this model with rotary embeddings applied to the _head_
 dimension, which is the default in xformers as of 09/01/2023. Since these models
 were trained, we have updated openlm to correctly apply the rotary embeddings to
 the sequence dimension (see
 [this issue](https://github.com/mlfoundations/open_lm/issues/4) and [this
 issue](https://github.com/facebookresearch/xformers/issues/841) for details).
-To evaluate these models, ensure you pass `--rotary-old` to the eval command.
+To evaluate these models, ensure you use the `m1b_neox_head_rotary.json` config.
```
Comment (achalddave) [inline README.md:148]: Should also be `--positional_embedding_type head_rotary`

File: eval/in_memory_hf_eval.yaml (hunk 2 above)
Comment (jmercat) [reply eval/in_memory_hf_eval.yaml:33]: Not really I can revert that.

File: README.md (hunk 6 above)
Comment (jmercat) [reply README.md:148]: I should add that in the json file but m1b_neox_head_rotary does not exist... I can create it.

Comment (achalddave) [reply README.md:148]: We deleted the head_rotary configs now, and the rotary flag is always just done with the command line flags, so no need to add

File: open_lm/model.py (hunk 4 above)
Comment (jmercat) [reply open_lm/model.py:312]: old configs might also match old models that use head_rotary so I thought it'd be better to fail explicitly. Tell me if you still want me to put rotary as default.

File: eval/eval_openlm_ckpt.py (hunk 1 above)
Comment (jmercat) [reply eval/eval_openlm_ckpt.py:122]: Actually the second point should not be a problem as long as we load with the same positional embedding it might be a bug I'll check that

Comment (jmercat) [reply eval/eval_openlm_ckpt.py:122]: There was a bug in the create_params function that I miss-interpreted as a difference in state_dicts. This is solved.

File: open_lm/utils/verify_converted_llama.ipynb
Comment (jmercat) [inline open_lm/utils/verify_converted_llama.ipynb:1]: Do we actually want this file there? I cannot run it and many paths are hard-coded in different places... Also I'm not a big user of notebooks and git seems to track some run results which we probably don't want.

Comment (sagadre) [reply open_lm/utils/verify_converted_llama.ipynb:1]: agreed. i think we maybe want a unit test that can be added in this or another PR. i’m personally ok with removing this file for now

File: open_lm/model.py (hunk 4 above)
Comment (sagadre) [reply open_lm/model.py]: nice!

**Review by sagadre — APPROVED** [review]
lgtm! still think we can consider removing the ipython, but no strong feelings

**Comment by jmercat** [comment]
> lgtm! still think we can consider removing the ipython, but no strong feelings

Great then I think it can be merged. The ipython can be removed and replaced by a test in a separate PR if desired. I was just mentioning it there because I didn’t test it and what I did should have an impact on it (for example adding the field state_dict in the saved weights and also a different default saving location than the one hard-coded in it...)

**Comment by sagadre** [comment]
sounds good! will merge and create an issue for a llama2 conversion unit test---thanks @jmercat!
